### DISCUSSION TITLE
Garbage Collection

### DISCUSSION THREAD

**Discussion by donalffons** [donalffons/opencascade.js#79] [Ideas]
I just came across [EigenJS](https://github.com/BertrandBev/eigen-js). They are using Emscripten + Embind to expose Eigen to JavaScript + WASM. In their README, they describe a potentially interesting approach to improve memory management.

**Comment by donalffons** [comment]
After thinking about this some more, I think that `Finalization Registry` might be the way to go. Here is a [HTTP 203 video](https://youtu.be/uygxJ8Wxotc?t=1058) on the topic. Would be interesting to do some prototyping.

**Reply by donalffons** [reply]
This is also mentioned in the [Emscripten documentation](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management). It would be worth a look if a [custom smart pointer](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#custom-smart-pointers) could be implemented for `opencascade::handle`.

**Reply by sgenoud** [reply]
Thanks for the tip on the `Finalization Registry` - this will make replicad object management so much easier!

**Reply by donalffons** [reply]
Yeah, it sounds pretty neat, doesn't it?! I'm not sure about browser support and reliability of that feature though. Getting this to work consistently for all wasm-interfacing code would be fantastic. Would be great to hear your experience after you had a chance of doing some prototyping with it.

**Reply by sgenoud** [reply]
I have played a bit with it today, and it seems to work fine with Chrome (which does not say much about it). If it does not work, it should not be a big issue (i.e. manual management is still possible).

On my end, I like having it at the replicad level - this feels like it belongs well to the adaptation layer between the very C++ interface of occt to something more JS / TS - like. But I would not be unhappy being able to just ignore memory management alltogether.

**Reply by sgenoud** [reply]
If someone wants to have a look, my implementation of the garbage collection using the `FinalizationRegistry` is [here](https://github.com/sgenoud/replicad/blob/main/packages/replicad/src/register.ts)

**Reply by donalffons** [reply]
Wow, that's great, thanks for sharing! I was confused for a second because I didn't see `Weak` types used anywhere - but then I realized that that's not even necessary for this to work. The implementation really doesn't look too difficult.
How did you test that? The "Collect Garbage" button in Chrome's Devtools? Did you have a chance to test that on Firefox / Safari?

**Reply by sgenoud** [reply]
> How did you test that? The "Collect Garbage" button in Chrome's Devtools?

Mostly yes, but depending on what I was doing it triggered automatically as well.

 > Did you have a chance to test that on Firefox / Safari?

Yes, and it also triggered automatically (depending on the model I was building as well).

I did not try to stress test it though, and I imagine that if you want to run something heavy it might be better to handle memory manually. But if this is your use case, running it in the browser might not be the best approach 😉 .

Ok, to show off I created [this small test](https://studio.replicad.xyz/share/https%3A%2F%2Fgist.githubusercontent.com%2Fsgenoud%2F9f65680ab31c1f0be3b73dd7e83b757d%2Fraw%2F05f7c53108c5bbab578b5a6133e415bc1fbfa930%2Ftest-gc.js), but for some reason it does not seem to work on Safari in this situation.

**Reply by donalffons** [reply]
Nice. I just also tested that site with a few devices (some with BrowserStack)
- Works with recent versions of Chrome + Firefox on my Laptop. Both browser show "deleting by gc" messages in the console.
- I tested that page on an iPhone 13, iOS 15.3, Safari - the page works but no logs appear about garbage collection.
- Works fine on a Galaxy S22, Android 12, Chrome

Great to see that working :slightly_smiling_face:.

**Reply by sgenoud** [reply]
Done some more experimenting with Safari, the gc seems to work, but not trigger with the gc button or as quickly as the other browsers 🤷 (especially with this particular test). It might be related to the particular test or some Safari things that I do not understand.

Additionally to the `Wrapper` class, one pattern that I found useful was to register objects to the functions they are used - they are cleaned when the function goes out of scope. I have used that pattern [here](https://github.com/sgenoud/replicad/blob/3e1a7c798cef4db158c0b1e67c2228c9faaf039c/packages/replicad/src/finders.ts#L150-L160).
